internals/overlord/logstate/loki.go

```diff
@@ -0,0 +1,157 @@
+)
+
+const (
+	maxLokiRequestSize = 10
+	lokiRequestTimeout = 10 * time.Second
+)
+
+type lokiClient struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
+}
+
+func newLokiClient(target *plan.LogTarget) logClient {
+	return &lokiClient{
+		target:     target,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: http.Client{Timeout: lokiRequestTimeout},
+	}
+}
+
+func (c *lokiClient) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], asLokiEntry(entry))
+	c.numEntries++
+	if c.numEntries >= maxLokiRequestSize {
+		return c.Flush(ctx)
+	}
+	return nil
+}
+
+func asLokiEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *lokiClient) Flush(ctx context.Context) error {
+	if c.numEntries == 0 {
+		return nil // no-op
+	}
+	defer c.emptyBuffer()
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	// Check response status code to see if it was successful
+	if code := resp.StatusCode; code < 200 || code >= 300 {
+		// Request to Loki failed
+		b, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+		if err != nil {
+			b = append(b, []byte("//couldn't read response body: ")...)
+			b = append(b, []byte(err.Error())...)
+		}
+		return fmt.Errorf("cannot send logs to Loki (HTTP %d), response body:\n%s", resp.StatusCode, b)
+	}
+
+	return nil
+}
+
+func (c *lokiClient) buildRequest() lokiRequest {
+	// Sort keys to guarantee deterministic output
+	var services []string
+	for svc, entries := range c.entries {
+		if len(entries) == 0 {
+			continue
+		}
+		services = append(services, svc)
+	}
+	sort.Strings(services)
+
+	var req lokiRequest
+	for _, service := range services {
+		entries := c.entries[service]
+		stream := lokiStream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+func (c *lokiClient) emptyBuffer() {
+	for svc := range c.entries {
+		c.entries[svc] = c.entries[svc][:0]
+	}
+	c.numEntries = 0
+}
+
+type lokiRequest struct {
```

Comment:
Yes, we need to call it `lokiClient` to prevent collision with other types of client in the `logstate` package.

You're not the first person to suggest moving this into its own package, and I think it makes a lot of sense, for the reasons you list. It would also reduce clutter in the `logstate` package, as we may add many more types of client in the future.

I think the best place to put it is a subpackage of `logstate`, i.e. `github.com/canonical/pebble/internals/overlord/logstate/loki`.